# Data generating distribution and convergence of linear regression.

(a) Generate a synthetic dataset with the following data generating process: $$Y \approx \alpha + \beta X + \epsilon,$$

where $X$ is drawn from a normal distribution with mean 168 and standard deviation 30, and $\epsilon$ is drawn from a normal distribution with mean 0 and standard deviation 20. Let $\alpha = 20$, and $\beta = 0.5$.

[Hint: use function 'numpy.random.normal', and $X$ and $Y$ should be a column vector. Think carefully what the shape of $X$ and epsilon should be when generating their values.]

In [11]:
import numpy as np

alpha = 20
beta = 0.5

def generate_X(n):
    X = np.random.normal(168, 30, (n, 1))
    return X

def generate_e(n):
    epsilon = np.random.normal(0, 20, (n, 1))
    return epsilon

def generate_Y(X, e):
    Y = alpha + beta * X + e
    return Y


(b) Run 5 linear regressions with sample sizes $n = 10^2, 10^3, 10^4, 10^5, 10^6$, respectively. Report the coefficients and $R^2$ of each run.

In [44]:
from sklearn.linear_model import LinearRegression

n = [10**2, 10**3, 10**4, 10**5, 10**6]

for size in n:
    X = generate_X(size)
    e = generate_e(size)
    Y = generate_Y(X, e)

    model = LinearRegression()
    model.fit(X, Y)
    r_sq = model.score(X, Y)

    print(f"Intercept (alpha) for n = {size}: {model.intercept_[0]}")
    print(f"Slope (beta) for n = {size}: {model.coef_[0, 0]}")
    print(f"R^2 for n = {size}: {r_sq}", "\n")

Intercept (alpha) for n = 100: 39.885653433248805
Slope (beta) for n = 100: 0.39519520121561247
R^2 for n = 100: 0.27870415365407675 

Intercept (alpha) for n = 1000: 23.544502803575156
Slope (beta) for n = 1000: 0.4796945780715412
R^2 for n = 1000: 0.313539331554855 

Intercept (alpha) for n = 10000: 19.741393281822226
Slope (beta) for n = 10000: 0.4995103064578878
R^2 for n = 10000: 0.36308134717612783 

Intercept (alpha) for n = 100000: 19.43711808955416
Slope (beta) for n = 100000: 0.5034787101657434
R^2 for n = 100000: 0.3647606118309439 

Intercept (alpha) for n = 1000000: 20.060102107902736
Slope (beta) for n = 1000000: 0.49972345074724217
R^2 for n = 1000000: 0.36005520265328983 



(c) What do you observe in part (b)? Do the coefficients of your linear regressions converge to the $\alpha$ and $\beta$? Does your $R^2$ converge to 0? If not, what number does it converge to?

The coefficients of my linear regression do indeed converge to the $\alpha$ and $\beta$ values of 20 and 0.5 respectively. However, the $R^2$ does not converge to zero, and instead converges to 0.36. This is because of the irreducible noise from $\epsilon$, so we see that the R^2 convergest to the following:

$var(Y) = \beta^2 var(X) + var(\epsilon) = 0.25(900) + 400 = 625$

$R^2 = 1 - \frac{var(\epsilon)}{var(Y)} = 1 - \frac{400}{625} = 1 - 0.64 = 0.36$

(d) Recall that the analytical form of $R^2$ is given by $$R^2 = 1 − \frac{\sum_{i=1}^n (y^{(i)} − \hat{y}^{(i)})^2}{\sum_{i=1}^n (y^{(i)} − \overline{y})^2},$$ 

where $\overline{y} = \sum_{i=1}^n \frac{1}{n}y^{(i)}$ is the empirical mean of the labels. Recall that the sample variance of a random variable $X$ is calculated by $var(X) = \frac{1}{n} \sum_{i=1}^n (x^{(i)} − \overline{x})^2$, where $\overline{x}$ is the sample mean. Assuming that the coefficient of the linear regression converges to those in true data generating distribution, establish that when the data generation is linear, $R^2$ of a linear regression converge to $1 − \frac{var(\epsilon)}{var(Y)}$. 

We multiply the numerator and demoninator by $\frac{1}{n}$, and have that: $$R^2 = 1 - \frac{\frac{1}{n}\sum_{i=1}^n (y^{(i)} - \hat y^{(i)})^2}{\frac{1}{n}\sum_{i=1}^n (y^{(i)} - \bar y)^2}$$

We see that the denominator is the sample variance of $Y$ as the problem points out. And then we can expand the numerator by substituting in $y^{(i)} = \alpha + \beta x^{(i)} + \epsilon^{(i)}$ and $\hat y^{(i)} = \hat \alpha + \hat \beta x^{(i)}$: $$R^2 = 1 - \frac{\frac{1}{n}\sum_{i=1}^n ((\alpha - \hat \alpha) + (\beta - \hat \beta)x^{(i)} + \epsilon^{(i)})^2}{var(Y)}$$

Since we showed in part (c) above that the coefficients of the linear regression converge to $\alpha$ and $\beta$, we have that each cross term in the numerator is either a coefficient error or a product of two of them, multiplied by a sample average like $\frac{1}{n} \sum_{i=1}^n x^{(i)}, \frac{1}{n} \sum_{i=1}^n (x^{(i)})^2,$ etc., so by the LLN those averages converge to a finite value that is then multiplied by the coefficient error that converges to 0. So all in all, every term vanishes except for: $$\frac{1}{n} \sum_{i=1}^n (\epsilon^{(i)})^2$$ This is just $\mathbb{E}[\epsilon^2] = var(\epsilon)$. So we finally arrive at the end result which is that when the data generation is linear, $R^2$ of a linear regresion converges to: $$1 - \frac{var(\epsilon)}{var(Y)}$$

(e)[Bonus point +2] Show that the coefficients of the linear regression converges to $\alpha$ and $\beta$ in the data generating process.

(f) In the example above, you have established that if 1) the data generating process is linear and 2) we fit the data using linear regression, then the R^2
converges to $1 − \frac{var(\epsilon)}{var(Y)}$. Given this information, do you think there exists another model such that you obtain a better (higher) $R^2$ asymptotically (when the data that you get goes to infinity)? If so, show such an example. If not, please argue why. (You do not need to show this mathematically.)

No, because given only $X$, no model can achieve a higher asymptotic $R^2$ than $1 − \frac{var(\epsilon)}{var(Y)} = 0.36$ on new data.

Since $Y = \alpha + \beta X + \epsilon$, with $\epsilon$ independent of $X$ and $\mathbb{E}[\epsilon] = 0$, the best possible predictor of $Y$ from $X$ under squared loss is $\mathbb{E}[Y \mid X] = \alpha + \beta X$. For any model $f$,

$$E[(Y - f(X))^2] = \text{var}(\epsilon) + E[(\alpha + \beta X - f(X))^2] \;\ge\; \text{var}(\epsilon).$$

The $\text{var}(\epsilon)$ term is irreducible because $\epsilon$ carries no information about $X$, so no function of $X$ can predict it. Linear regression converges to $\alpha + \beta X$ as part e states, which drives the second term to zero and attains the lower bound. Any other model can at best tie it. A more flexible model, such as a high-degree polynomial or 1-nearest neighbor, can reach a higher in-sample $R^2$, up to 1, by interpolating the training points. But it does this by fitting the noise $\epsilon$, so its $R^2$ on new data falls below 0.36. That is overfitting, not a better model. As such, $R^2$ could exceed 0.36 only if we observed additional features correlated with ε, which changes the problem.

(g) Based on this exercise, what are some important characteristics of the data generating distribution necessary for the presence of a “good” model?

Based on this exercise, I think a good model depends on several properties of the data generating distribution. Most importantly, it needs a high signal-to-noise ratio, meaning the best achievable $R^2$ is $1 − \frac{\text{var}(\epsilon)}{\text{var}(Y)}$, so a model can explain $Y$ well only if most of $Y$'s variation comes from the features rather than from noise. Here, 64% of $\text{var}(Y)$ is noise, so even the true model reaches only $R^2 = 0.36$. The features must also capture the factors that actually drive $Y$, since anything unobserved becomes part of the irreducible error. For the coefficients to be estimated correctly, the noise must be uncorrelated with the features. Otherwise, $\hat{\beta}$ converges to $\beta + \frac{\text{Cov}(X, \epsilon)}{\text{var}(X)}$ rather than $\beta$. $X$ must also vary enough ($var(X) > 0$) for its effect on $Y$ to be identified. In addition, the model class must be able to represent $\mathbb{E}[Y \mid X]$, as linear regression can here. Finally, the samples should be i.i.d. from a stable distribution, in sufficient quantity for the law of large numbers to apply. A low $R^2$ therefore does not necessarily mean the model is bad, it may just be noisy data.